Add an RSSI sensor to the LIFX integration

Djelibeybi opened the pull request:
## Proposed change

Add an RSSI sensor to the LIFX integration which is classed as diagnostic and disabled by default. 

Also switch all sensor entities to use a sensor-specific coordinator that updates way less often than the default coordinator. This means some values (like RSSI or infrared brightness) are updated less often, but only affects attributes that are not visible, i.e. things like color and power still update every 10 seconds.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/24551

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [X] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Djelibeybi committed:
Add an RSSI sensor to the LIFX integration

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi committed:
Add an RSSI sensor to the LIFX integration

Signed-off-by: Avi Miller <me@dje.li>

homeassistant commented:
Hey there @bdraco, mind taking a look at this pull request as it has been labeled with an integration (`lifx`) you are listed as a [code owner](https://github.com/home-assistant/core/blob/dev/CODEOWNERS#L624) for? Thanks!

bdraco commented on homeassistant/components/lifx/sensor.py:
```diff
@@ -0,0 +1,74 @@
+    @callback
+    async def async_added_to_hass(self) -> None:
+        """Get the coordinator to update RSSI every minute."""
+
+        self.async_on_remove(
+            async_track_time_interval(
+                self.hass, self.coordinator.async_update_rssi, SCAN_INTERVAL
+            )
+        )
```
If you set `_attr_should_poll` to `True` and move this to an `async_update` you can drop `async_added_to_hass` but either way its a bit messy because you'll have to implement exception handling.

Djelibeybi replied:
This seemed like the less convoluted option and is way less complicated than the last time I attempted this.

Djelibeybi replied:
Also, I didn't want this to poll as often as the `DataUpdateCoordinator` updates but I also didn't want to create a whole new coordinator for one sensor.

bdraco replied:
Yeah it's definitely a sea of not so great options.

I would probably still do another coordinator since it avoids a lot of the boiler plate

Djelibeybi replied:
That will have to wait until later in the week, then.

Djelibeybi committed:
Sensor entities now use their own coordinator that syncs less often

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi committed:
Sensor entities now use their own coordinator that syncs less often

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi commented:
@bdraco I'm not sure if this implementation of the sensor coordinator is madness or genius, but it works. :)

Djelibeybi commented:
I'll open a doc PR once we agree on the manner in which this is implemented so I can document the different update periods, etc.

Djelibeybi commented:
@bdraco gentle bump. :)

epenet committed:
Add websocket type hints in entity_registry (#80657)

* Add websocket type hints in entity_registry

* Adjust websocket_list_entities

* Fix update

* Fix websocket_update_entity

* Apply suggestion

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Apply suggestion

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bramkragten committed:
Update frontend to 20221020.0 (#80661)

epenet committed:
Improve type hint in entity_component (#80596)

Imrpove type hint in entity_component

avishorp committed:
Add mode control for Modbus climate entities (#73906)

* Add support for Modbus HVAC control registers

epenet committed:
Add websocket type hints in components (#80654)

* Add websocket type hints in components

* Adjust

* Apply suggestion

Co-authored-by: Marc Mueller <30130371+cdce8p@users.noreply.github.com>

Co-authored-by: Marc Mueller <30130371+cdce8p@users.noreply.github.com>

thecode committed:
Bump aioshelly to 4.0.0 (#80423)

* Bump aioshelly to 4.0.0

* Remove leftover

* Fix number platform

* Set last_update_success to false upon failure in number and climate

* Set last_update_success upon failurie in entity

riokuu committed:
Add blebox binary_sensor platform (#79535)

* Add binary_sensor platform, with test.

* Applied suggestions by @epenet

* refactor: as @epenet suggested, passing entity_description to init

* Update homeassistant/components/blebox/binary_sensor.py

@epenet suggestion, moved refactored logic of create_blebox_entities into BleBoxBinarySensorEntity init

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

* refactor: as @epenet class selector and entity creation moved to binary_sensor

* refactor: list comprehension in entity list setup in binary sensor

* Update homeassistant/components/blebox/binary_sensor.py

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

epenet committed:
Use US_CUSTOMARY_SYSTEM in tests (#80658)

* Use US_CUSTOMARY_SYSTEM in tests

* Don't update test_unit_system

riokuu committed:
Refactor blebox sensors (#80671)

refactor: sensors, entity description pick moved to async_setup_entry, removed redundant dataclass

epenet committed:
Add websocket type hints in lovelace (#80537)

epenet committed:
Adjust device classes in tasmota (#79282)

* Adjust device classes in tasmota

* Remove incorrect device class

bdraco committed:
Speed up restart when bluetooth adapter is in a failed state (#80640)

eifinger committed:
Remove deprecated yaml import for waze_travel_time (#80669)

Remove deprecated yaml import

Signed-off-by: Kevin Stillhammer <kevin.stillhammer@gmail.com>

Signed-off-by: Kevin Stillhammer <kevin.stillhammer@gmail.com>

balloob committed:
Remove balloob as code owner Ring (#80680)

frenck committed:
Pin uamqp==1.6.0 (#80678)

balloob committed:
Fix custom components not working with integration descriptions (#80686)

Kane610 committed:
Simplify mocking UniFi websocket (#80652)

* Simplify mocking UniFi websocket

* Fix one more

bdraco committed:
Fix bluetooth calls from automations in esphome (#80683)

bdraco committed:
Defer bluetooth scanner watchdog restart if one is already in progress (#80679)

luar123 committed:
Bump snapcast to 2.3.0 (#80688)

Commit:
[ci skip] Translation update

bdraco committed:
Bump zeroconf to 0.39.2 (#80699)

python 3.11 support & performance improvements

changelog: https://github.com/jstasiak/python-zeroconf/compare/0.39.1...0.39.2

ctalkington committed:
Add diagnostics to jellyfin (#80651)

bachya committed:
Replace custom OpenUV data object with coordinators (#80705)

* Replace custom OpenUV data object with coordinators

* Typing

* Code review

eifinger committed:
Move default option handling to config_flow for google_travel_time (#80607)

Move default option handling to config_flow

cgtobi committed:
Bump pyatmo to 7.2.0 (#80698)

cgtobi committed:
Fix error when setting Netatmo climate preset (#80700)

frenck committed:
Integrations v2.1: Virtual integrations (#80613)

Hadatko committed:
Fix nextcloud 'ncm' referenced before assignment (#80711)

* #80673 Fix 'ncm' referenced before assignmen

 UnboundLocalError: local variable 'ncm' referenced before assignmen

Signed-off-by: Dusan Cervenka <cervenka.dusan@gmail.com>

* #80673 changes based on review

Signed-off-by: Dusan Cervenka <cervenka.dusan@gmail.com>

* Changes made based on review

Signed-off-by: Dusan Cervenka <cervenka.dusan@gmail.com>

Signed-off-by: Dusan Cervenka <cervenka.dusan@gmail.com>

bouwew committed:
Add entity_registry_enabled_default and missing EntityCategories in Plugwise (#80629)

renegaderyu committed:
Fix ZeroDivisionError for Fritz!Smarthome electric current sensor (#80682)

Fixes ZeroDivisionError for fritzbox sensor. Fixes #80618.

allenporter committed:
Improve calendar trigger test quality (#79451)

Improve calendar test quality

allenporter committed:
Streamline calendar dataclass API/attribute conversions (#79598)

* Streamline calendar dataclass API/attribute conversions

* Fix attribute conversions

Shutgun committed:
Add iot_standards to devolo brand (#80332)

jbouwh committed:
Improve typing hints MQTT __init__ (#80674)

* Improve typing __init__

* Follow up suggestions

fredrike committed:
Add Bluesound unique id (#80559)

* add unique_id & device_info

* remove broken image

* use a combination of mac and port (for multizone devices)

* use the typed dataclass

Co-authored-by: Mick Vleeshouwer <mick@imick.nl>

* Don't use get

* fix device_info

* remove device_info

Co-authored-by: Mick Vleeshouwer <mick@imick.nl>

frenck committed:
Update home-assistant/wheels to 2022.10.1 (#80723)

frenck committed:
Update sentry-sdk to 1.10.0 (#80721)

pvizeli committed:
Remove building wheels for face detection (#80728)

tymm committed:
Bump simplepush to 2.1.1 (#80608)

* Update to new library version

* Remove test for removed send_encrypted function

* Bump simplepush to 2.1.1

uvjustin committed:
Use empty_moov in stream recorder (#80726)

pnbruckner committed:
Increase life360 timeout (#80692)

frenck committed:
Small typing improvements for Plugwise (#80722)

allenporter committed:
Reduce unnecessary alarm firing to speed up calendar trigger test (#80732)

bdraco committed:
Log invalid messages instead of raising in system_log (#80645)

eifinger committed:
Move default options to config_flow for waze_travel_time (#80681)

Move default options to config_flow

Signed-off-by: Kevin Stillhammer <kevin.stillhammer@gmail.com>

Signed-off-by: Kevin Stillhammer <kevin.stillhammer@gmail.com>

engrbm87 committed:
Retry setup in case of empty response from Pushover api (#80602)

Retry setup in case of empty response

starkillerOG committed:
Remove Xiaomi Miio YAML import (#78995)

* Deprecate YAML import

* Add logging for unexpected errors

* remove unused import

* fix tests

* unused import

* fix tests

* fix snake_case

* Do not add to standard key string

bdraco committed:
Fix zha LogEntry call (#80737)

The signature for LogEntry changed in #80645 to drop the
unused argument

bdraco committed:
Remove system_log missing format arg test (#80739)

javicalle committed:
Add `_TZE200_kds0pmmv` to ZHA `ZONNSMARTThermostat` (#80746)

Add `_TZE200_kds0pmmv` to `ZONNSMARTThermostat`

Related to: https://github.com/zigpy/zha-device-handlers/pull/1843

bdraco committed:
Add homekit notifications to the list of allowed apple bluetooth start bytes (#80733)

bachya committed:
Remove redundant `async_update` for OpenUV entities (#80735)

bdraco committed:
Bump qingping-ble to 0.8.2 to handle door left open (#80748)

fixes #78439

starkillerOG committed:
Fix Goodwe spinlock (#80624)

* fix spinlock

* Add debug log

* fix styling

rklomp committed:
Bump pysma to 0.7.1 (#80601)

* Bump pysma to 0.7.0

* Bump pysma to 0.7.1

* Fix test

kevdliu committed:
Load ecobee notify platform via discovery (#78558)

* Fix ecobee notify platform KeyError

* set up notify platform via discovery

* address comments

* fix isort

* Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

cdce8p committed:
Update pylint to 2.15.5 (#80759)

balloob committed:
Conditionally include config flow and iot_class when relevant (#80756)

balloob committed:
Add ultraloq virtual integration (#80755)

bramkragten committed:
Update frontend to 20221021.0 (#80751)

Commit:
[ci skip] Translation update

balloob committed:
Remove unused keys from ultraloq (#80762)

TheJulianJES committed:
Add "power on state" config entity for Tuya plugs to ZHA (#80486)

* Add "power on state" config entity for Tuya plugs to ZHA

* Remove TS004F from models, as it doesn't support power-on-state

TheJulianJES committed:
Add "power outage memory" config entity to Xiaomi EU plugs to ZHA (#80444)

TheJulianJES committed:
Add Philips Hue motion sensor config entities to ZHA (#79923)

* Add Hue trigger LED config switch

* Add Hue sensitivity config select

* Use existing consts for channel names

* Add friendly names to config entities

* Follow HA capitalization conventions

* Move Hue motion sensor check to a helper method

* Move helper method to a new helpers file in channels folder

balloob committed:
Update pending message error (#80763)

kpine committed:
Update zwave_js.refresh_value service example (#80764)

thecode committed:
Fix Shelly entry unload and add tests for init (#80760)

bieniu committed:
Remove myself from Xiaomi Miio code owner list (#80768)

bieniu committed:
Add `integration_type` field for a few integrations (#80767)

epenet committed:
Add myself as code owner to util and const (#80664)

Shutgun committed:
Add integration_type to devolo integrations (#80695)

uvjustin committed:
Bump owntone requirements (#80552)

onpremcloudguy committed:
Extend roomba mac range in manifest (#80714)

Co-authored-by: Franck Nijhof <git@frenck.dev>

bdraco committed:
Bump bleak-retry-connector to 2.3.2 (#80790)

Commit:
[ci skip] Translation update

jjlawren committed:
Bump soco to 0.28.1 (#80792)

thecode committed:
Bump aioshelly to 4.1.0 (#80795)

bdraco committed:
Bump aiohomekit to 2.2.0 (#80798)

bdraco committed:
Log bluetooth advertisement before firing bleak callbacks (#80800)

The debug log was confusing because the bleak callbacks
were firing before we were logging the advertisements

frenck committed:
Update psutil to 5.9.3 (#80775)

bouwew committed:
Bump plugwise to v0.25.3 (#80782)

kingy444 committed:
Refactor Powerview sensor to dataclass (#80536)

Djelibeybi committed:
Add themes for LIFX multi-zone devices via a new select entity (#80067)

Kane610 committed:
Refactor UniFi upgrade entities (#80752)

* Refactor UniFi upgrade entities

* Enable type check for UniFi update platform

fredrike committed:
Bump pydaikin to 2.8.0 (#80823)

Kane610 committed:
Refactor UniFi outlet switches (#80738)

* Rewrite UniFi outlet switches

* Bump aiounifi to v41

* Remove devices from items_added input

ctalkington committed:
Add integration_type to ipp and roku (#80831)

Kane610 committed:
Improve UniFi PoE unique ID (#80740)

* Improve UniFi PoE unique ID

* Set available property on creation

bdraco committed:
Prevent HomeKit Controller BLE connect retries from blocking startup (#80827)

ctalkington committed:
Add integration_type to jellyfin (#80832)

frenck committed:
Set integration type on codeowned integrations @frenck (#80830)

gjohansson-ST committed:
Add entity service - Set Full AC state to Sensibo (#80820)

Co-authored-by: J. Nick Koston <nick@koston.org>

gjohansson-ST committed:
Minor cleanup Sensibo (#80835)

bdraco committed:
Small lifx fixes (#80828)

danaues committed:
Add dynamic generation of device triggers from keypad buttons (#80797)

Co-authored-by: J. Nick Koston <nick@koston.org>

natekspencer committed:
Bump pylitterbot to 2022.10.2 (#80836)

gjohansson-ST committed:
Add Sensibo Climate React (#78221)

gjohansson-ST committed:
Fix temperature unit in sensor for Sensibo (#80843)

Kane610 committed:
Add presence duration number (#79498)

Kane610 committed:
Refactor UniFi DPI switch entities (#80761)

* Refactor UniFi DPI switch entities

* Remove dpi presence from items_added

danaues committed:
Fix Lutron Caseta area names by ignoring root area during area name retrieval (#80576)

jloehr committed:
Add ZHA StartUpColorTemperature configuration entity (#80853)

* Add ZHA start up color temperature entity

* Use device reported min max values

* Add color number test

* Fix code style

PeteRager committed:
Fix oncue data unavailable when genset disconnected (#80668)

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco committed:
Try again to populate HKC BLE accessory state after startup (#80854)

danaues committed:
Add Lutron Caseta zeroconf discovery for RA3/HWQSX (#80852)

bdraco committed:
Fix whitespace in oncue manifest (#80859)

G-Two committed:
Code quality update for Subaru sensors (#79482)

* Use distance device class for sensors

* Change sensor name casing and unique_id

* Migrate sensor entity unique_id

* Match title-cased unique_id when migrating

* Remove unneeded regex to find '_' delimited id suffix

* Incorporate PR review comments

* Add check to prevent extra odometer entity migration

oyvindwe committed:
Support Nobø Switch as temperature sensor (#78480)

* Expose Nobø Switch as temperatur sensor.

* - Review
- Hub may report current temperature as None

* Avoid update during entity addition, and fix race condition

* Update pynobo to 1.6.0
Use new method to fix potential race condition.

* Use generator expressions

GitHub-Action committed:
[ci skip] Translation update

bdraco committed:
Bump aiohomekit to 2.2.2 (#80857)

* Bump aiohomekit to 2.2.2

changelog: https://github.com/Jc2k/aiohomekit/compare/2.2.1...2.2.2

* ci bump

ol-iver committed:
Update denonavr to version 0.10.12 (#80861)

Djelibeybi committed:
Add device-specific diagnostics to the LIFX integration (#79964)

allenporter committed:
Bump gcal_sync to 2.2.0 (#80863)

Quentame committed:
Add Freebox button to mark calls as read (#80609)

* Add Freebox button to mark calls as read

* Bump Freebox to 1.0.1

* Fix black/flake8

* Add entity_category + fix reboot button name

* Add has_entity_name to reboot

* Remove 'missed' as it put all calls as read

* base unique_id on key and not name

* unique_id base on key later with migration step

* Keep the same name

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

dependabot[bot] committed:
Bump actions/upload-artifact from 3.1.0 to 3.1.1 (#80867)

Bumps [actions/upload-artifact](https://github.com/actions/upload-artifact) from 3.1.0 to 3.1.1.
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](https://github.com/actions/upload-artifact/compare/v3.1.0...v3.1.1)

---
updated-dependencies:
- dependency-name: actions/upload-artifact
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

jbouwh committed:
Move advanced MQTT options to entry (#79351)

* Move advanced broker settings to entry

* Add repair issue for deprecated settings

* Split CONFIG_SCHEMA

* Do not store certificate UI flags in entry

* Keep entered password in next dialog

* Do not process yaml config in flow

* Correct typo

epenet committed:
Remove precipitation note from length units (#80677)

bieniu committed:
Add `update` platform to MQTT integration (#80659)

epenet committed:
CI: Fix partial coverage (#80877)

* CI: Split coverage job

* Don't wait for pytest-mariadb

elupus committed:
Allow hostname for nibe heatpump (#80793)

Allow hostname for nibe

epenet committed:
Use DataUpdateCoordinator in scrape (#80593)

* Add DataUpdateCoordinator to scrape

* Fix tests

jbouwh committed:
Add typing hints for MQTT mixins (#80702)

* Add typing hints for MQTT mixins

* Follow up comments

* config_entry is always set

* typing discovery_data - substate None assignment

* Rename `config[CONF_DEVICE]` -> specifications

epenet committed:
Add unique_id to scrape (#80581)

* Add unique_id to scrape

* Add tests

epenet committed:
Adjust pylint for IMPERIAL_SYSTEM deprecation (#80874)

* Adjust pylint for IMPERIAL_SYSTEM deprecation

* Use correct location

* Adjust components

bdraco reviewed (changes requested):
Please resolve conflicts

emontnemery committed:
Add suggested_unit_of_measurement attribute to sensors (#80638)

* Add suggested_unit_of_measurement attribute to sensors

* Lazy calculation of initial entity options

* Add type alias for entity options

* Small tweak

* Add tests

* Store suggested_unit_of_measurement in its own option key

* Adapt to renaming of IMPERIAL_SYSTEM

* Fix rebase mistakes

* Apply suggestions from code review

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

elupus committed:
Add field descriptions for nibe heat pump (#80791)

bdraco committed:
Bump bleak-retry-connector to 2.4.0 (#80887)

bdraco committed:
Fix HKC exceptions during BLE startup not being caught (#80882)

bdraco committed:
Bump aiohomekit to 2.2.3 (#80891)

Jc2k committed:
Add support for Netatmo noise sensor to homekit_controller (#80889)

Co-authored-by: J. Nick Koston <nick@koston.org>

MartinHjelmare committed:
Refactor zwave_js add-on manager (#80883)

* Make addon slug an instance attribute

* Extract addon name and addon config

* Update docstrings

rikroe committed:
Add EMISSION_CHECK to BMW Connected Drive (#80819)

bdraco committed:
Bump aiohomekit to 2.2.4 (#80899)

Djelibeybi commented:
Whoops, I stuffed that up, didn't I? I'll submit a new PR later today/tomorrow when I get some spare cycles.
